### PR TITLE
Fixes 21079 issue, periodicity, solvify for piecewise

### LINKED ISSUE #21079: Solvify cannot solve Piecewise
```
>>> p = Piecewise((0, x < -1), (x**2, x <= 1), (log(x), True))
>>> solvify(p,x,S.Reals)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sayandip/sympy/sympy/solvers/solveset.py", line 2446, in solvify
    period = periodicity(f, symbol)
  File "/home/sayandip/sympy/sympy/calculus/util.py", line 516, in periodicity
    g_s = decompogen(f, symbol)
  File "/home/sayandip/sympy/sympy/solvers/decompogen.py", line 47, in decompogen
    result += [f.subs(f.args[0], symbol)] + decompogen(f.args[0], symbol)
  File "/home/sayandip/sympy/sympy/solvers/decompogen.py", line 37, in decompogen
    raise TypeError('expecting Expr but got: `%s`' % func_name(f))
TypeError: expecting Expr but got: `ExprCondPair`
>>> solveset(p,x,S.Reals)
Union(FiniteSet(0), Interval.open(-oo, -1))

```

### PR DESCRIPTION
**Pull request by MaqOwais**
`solvify` is now able to solve `piecewise` expressions.
Fixes  #21079 as it was giving an error due to a bug in `periodicity`.
Hence, periodicity for `piecewise` is fixed as `None` as the period of `piecewise` should also be `piecewise`.(which is an inconsistent return type)

```
>>> p = Piecewise((0, x < -10), (x**2 + 5*x - 6, x >= -9))
>>> solvify(p, x, S.Reals) 
        [-6, 1]

>>> periodicity(p, x) is None
        True
```
Closes & Fixes #21079

NO ENTRY

### CODE REVIEW

**Comment by MaqOwais** [comment]
ping @smichr 
I think this should be merged first!

**Comment by smichr** [comment]
Looks pretty good. Could we raise NotImplementedError for the case of Piecewise? e.g. `Piecewise((1, sin(x) > 1), (0, True))` has a well defined period of `2*pi` -- and there could be cases of more args, too.

**Comment by MaqOwais** [comment]
> Could we raise NotImplementedError for the case of Piecewise? 

Actually, I think there is no need for this, since if `solveset` gives a solution then `solvify` will be giving a solution for the same. 

```
>>> p = Piecewise((0, x < -10), (x**2 + 5*x - 6, x >= -9))
>>> solveset(p,x,Reals)
Union(FiniteSet(-6, 1), Interval.open(-oo, -10))
>>> solvify(p, x, S.Reals) 
        [-6, 1]

```
Also, maybe the solution may be not complete since it has neglected the interval part ` Interval.open(-oo, -10)` (but this gives a partial solution that is for sure).

**Comment by MaqOwais** [comment]
I think the incomplete solution will be covered with https://github.com/sympy/sympy/pull/21174 .

File: sympy/calculus/util.py [hunk 1]
```diff
@@ -511,6 +511,11 @@ def _check(orig_f, period):
             symbol not in n.free_symbols):
                 period = Abs(n / a.diff(symbol))
 
+    elif isinstance(f, Piecewise):
+        # Returning None, as the return type, period of the `piecewise`
+        # also should be a piecewise. (i.e. the return type is not favorable)
+        return None
```
Comment (smichr) [inline sympy/calculus/util.py:517]:
```suggestion
        pass  # not handling Piecewise yet
```

**Comment by smichr** [comment]
> solution may be not complete since it has neglected the interval part

Could we return non-Finite sets as relationals in a Piecewise when possible so
```python
>>> p = Piecewise((0, x < -10), (x**2 + 5*x - 6, x >= -9))
>>> solveset(p,x,Reals)
Union(FiniteSet(-6, 1), Interval.open(-oo, -10))
>>> solvify(p, x, S.Reals) 
[-6, 1, Piecewise((x, x < -10))]
```
or maybe as a Piecewise?
```python
Piecewise((x, Or(x < -10, Eq(x, -6), Eq(x, 1))))
```
This doesn't have to be part of this PR, however. (And I'm not sure it needs to be done.)

**Comment by MaqOwais** [comment]
> I think the incomplete solution will be covered with #21174 .

```
>>> p = Piecewise((0, x < -10), (x**2 + 5*x - 6, x >= -9))
>>> solveset(p,x,Reals)
Union(FiniteSet(-6, 1), Interval.open(-oo, -10))
>>> solvify(p, x, S.Reals) 
[-6, 1, Interval.open(-oo, -10)]
```
What about this? (And this will be covered in #21174)

**Comment by smichr** [comment]
> What about this?

I don't know enough about the expectations of `solvify`; it seems reasonable that a user would want all the solution information. A format that might work is `[a,b,c,[sets]]` where `a,b,c` refer to particular solutions while `sets` refers to things that represent infinite solutions. But perhaps it is too late for such API changes. You'll have to consult with others.

**Comment by MaqOwais** [comment]
> I don't know enough about the expectations of solvify; it seems reasonable that a user would want all the solution information. 

No, it's just a try to convert the solveset result into the `solve` API format.

> You'll have to consult with others.

I think @oscarbenjamin has to comment on this step.

> A format that might work is [a,b,c,[sets]] where a,b,c refer to particular solutions while sets refers to things that represent infinite solutions.

When the changes from #21174 merge into this branch then `[a,b,c,interval]` like `[-6, 1, Interval.open(-oo, -10)]` format is observed.

**Comment by smichr** [comment]
ok, this is in.

**Pull request merged**